### ISSUE TITLE
What about using Mocket as core library?

### ISSUE THREAD

**Issue by mindflayer** [h2non/pook#33]
Hi there, I am the author of `mocket` and I think our projects are really similar and well connected.
What about joining forces? I think `mocket` is still a great idea, but currently I am the only one still working on it (we were two friends/colleagues at the time we presented it at EuroPython 2013, and another colleague helped us to fix some issues).
My idea is basically having `mocket` as `pook` engine.
What do you think about it?
Thanks in advance,
Giorgio

https://github.com/mocketize/python-mocket

**Comment by h2non** [comment]
Thanks for the proposal. `mocket` looks interesting.

While I appreciate and see good value in your library, when I'm implementing `pook` I made specific design decisions to make it high level, so no socket level programming is required, and therefore dealing with OSI layer protocol detail at all. 
IMO this provides more versatility, implementation simplicity, and considerably less coupling to networking APIs that are generally hard to support across Python versions.
Moreover, today we are more networking APIs than ever. Only in CPython we have traditional blocking or async I/O networking programming APIs, such as using `asyncio` in Python +3.4. 

You can see an example of what I'm exposing here in popular projects like HTTPretty, which does not provide Python 3 support at all due to the opposite design choice as Python.

As cons, `pook` is strongly coupled to HTTP libraries APIs, but that's a minor issue compared to the opposite scenario and it's clearly easier to maintain and scale anyway.

So basically, for all those reasons, I'd prefer keeping `pook` as it is today and preserve its original design strategy.

**Comment by h2non** [comment]
Worth saying that I'm very open to collaborate with you if you wanna improve `pook`.

And thank you for the suggestion, I appreciate that.

**Comment by mindflayer** [comment]
Please note that Mocket uses the same approach used by HTTPretty and provides the same functionalities on both Python 2 and Python 3 (we support even Python 2.6).

**Comment by mindflayer** [comment]
By the way, I understand your point even if I have the opposite standpoint, since I feel like mocking the socket is definitely easier than investing a lot of resources in maintaining the compatibility with all the main libraries.

**Comment by h2non** [comment]
Are you supporting asyncronous network I/O in order to mock HTTP traffic triggered by libraries such as `aiohttp`?

The time I vested mocking 3 HTTP libraries was just a couple of hours. For me it's definitively better, higher level and therefore less error prone compared to other approaches. Not having to parse raw HTTP protocol just makes it stronger by itself. Also, if you have to support socket streams, that would make things even more complex. pook uses an interceptor strategy which makes it higher level, so you only have to deal with high level HTTP library specific entities without caring about protocol level parsing or socket programming.

Just wanted to make clearify this. I think your solution is definitively interesting, but honestly I think It won't fit here.

**Comment by mindflayer** [comment]
Supporting async I/O is not a big deal and I have a working example. I will probably add the support soon.

**Comment by mindflayer** [comment]
Here it is: https://pypi.python.org/pypi/mocket/1.4.0

**Comment by h2non** [comment]
Nice, but I still can't see how that would work with asynchronous I/O networking based on `asyncio`. It would be nice having a working example for `aiohttp` client.

**Comment by mindflayer** [comment]
Of course the change has been shipped with proper tests. You will find them
in `tests35/`.

Il 17 dic 2016 2:29 AM, "Tomás Aparicio" <notifications@github.com> ha
scritto:

**Comment by h2non** [comment]
I think we can find an intermediate scenario where both parts are happy with and take advantage of both projects, which is the most relevant thing here :)

Since `pook` is very modular and was designed for extensibility, I think I can easily support a way to plug in a custom mock engine that would replace the built-in one. 
This would provide to the user a high level of customization.

In code terms, it may look like this:
````
pook.set_mock_engine(mocket)
```

Where `mocket` would be a callable object implementing a simple interface:
```python
MockEngine.__init__(self, pook_engine)
MockEngine.activate(self)
MockEngine.disable(self)
```

The interface is basically analog to how `pook` interceptors are implemented. We can leverage that interface for mock engines too.

Intercepted HTTP traffic by the mock engine should be normalized via a homogenus interface: `pook.Request`. This would be required in order to trigger the mocks matching engine.

I think this won't be hard to achieve. If we agree, I can provide `pook` API support, possibly this weekend.

Opinions? Do you see this easily viable from your side? @mindflayer

**Comment by h2non** [comment]
Good news: `pook` is now able to work with multiple mock engines that can be defined the user programmatically, as I described in the proposal above.

I think this provides strong freedom and versatility to package consumers in case that they want to use a custom HTTP mock engine, such as in your case with `mocket`.

For implementation details, please see: https://github.com/h2non/pook/commit/1ea37962a741fa23dadbf9d7c933f8484d640f89

Mock engine specific API documentation:
http://pook.readthedocs.io/en/latest/api.html#pook.MockEngine

This feature was introduced in `pook@0.1.7` and it's available via `PyPI` registry:
https://github.com/h2non/pook/blob/master/History.rst#v017--2016-12-18

If you have the chance to build an adapter for `mocket`, which I think should not be hard, please feel free to share your thoughts here.

If it works consistently, I'm open to add specific support for `mocket` as optional, activable capability in `pook` by default, also adding explicit support notes across documentation.

**Comment by mindflayer** [comment]
I'll have a look at it in the next days. Thank you.

**Comment by mindflayer** [comment]
I am looking into it right now, the big thing looks like creating a Mocket _Entry_ every time we define a pook._method_, what tells the engine when it's time to call the fake send[all] and when to use the real _socket_.

**Comment by h2non** [comment]
I'm not sure to follow. The mock engine is global, so once you define the one you want to use, it will apply to all the mock definitions afterward.

Can you share some code here?

**Comment by mindflayer** [comment]
`    Entry.single_register(
        Entry.GET,
        url_to_mock,
        body=json.dumps(response),
        headers={'content-type': 'application/json'}
    )`
This is the _mocket_ way to define an entry point to mock.
If you don't do this, the related call runs on the real _socket_.

**Comment by h2non** [comment]
I thought the idea about using `pook` with `mocket` was in order to take the advantage of its high-level API with expressive DSL.

From my perspective, you won't need to define mocks at `mocket` level (or any `mocket` specific abstraction for mock registration), you simply need to parse and intercept the outgoing HTTP request via socket stream, compose the `pook.Request` with the outgoing request, and finally trigger the mock engine matching phase, that would eventually give you the desired HTTP response to be mocked, or in case that no HTTP mock is matched, it would return nothing, where you should decide if failing notifying the user, or simply perform a real HTTP request (only if real networking mode is enabled).

Is this helpful for you?

**Comment by mindflayer** [comment]
Yes, the idea is taking advantage of its nice DSL.
But the only reason for having Mocket as its engine should be having the real client talking with the _MocketSocket_, instead of mocking it at client level like you do with your _interceptors_.

**Comment by mindflayer** [comment]
I see it possible since you are calling the _MockEngine_ here:
```
def mock(url=None, **kw):
    """
    Creates and register a new HTTP mock.
    Arguments:
        url (str): request URL to mock.
        **kw (mixed): variadic keyword arguments.
    Returns:
        pook.Mock: mock instance
    """
    return _engine.mock(url, **kw)
```

**Comment by mindflayer** [comment]
The only blocker I see is that _MockEngine_ does not inherit from _Engine_.

**Comment by h2non** [comment]
You don't need to inherit from `Engine`. This is intentional.  Currently used `Engine` instance in pook will be injected into the `MockEngine`. That's part of the minimum interface contract for custom `MockEngine`. See:
https://github.com/h2non/pook/blob/master/pook/mock_engine.py#L31-L32

**Comment by mindflayer** [comment]
Sorry, I edited my message while you were replying to me. I meant it does not implement that _mock_ method.

**Comment by h2non** [comment]
Why do you need that method at mock engine level?

**Comment by mindflayer** [comment]
Implementing _mock()_, or a method that hooks it, everything should run smoothly with _MocketEngine(MockEngine)_ under the hood.

**Comment by h2non** [comment]
I think in this context `activate()` method is the analog one. I'm still not sure why you want a `mock()` method since the mock definition is only a responsibility of `pook`.

There was an explicit concerns separation between `pook` engine and the underlying `mock` engine. The `mock` engine should only be responsible of activating/deactivating the mocking instrumentation and responsible of collecting outgoing HTTP traffic and triggering the mock matching engine in order to discover if an HTTP request should be mocked out or not. 

Perhaps I'm missing something, but the current mock engine is pretty self-explanatory IMO.

**Comment by h2non** [comment]
Also, you don't need to do any sort of inheritance for `MocketEngine`. 
If you implement the required interface contract everything would be fine:
https://github.com/h2non/pook/blob/master/pook/mock_engine.py#L9-L19

**Comment by mindflayer** [comment]
With _MocketEngine_ I mean my implementation of a _pook.MockEngine_.
I've already read the implementation, and almost all the rest of _pook_ implementation.
The point is: I do not see a way of having _Mocket_ as being an effective _pook.MockEngine_, since Mocket is based on the concept of having same Entry.

**Comment by mindflayer** [comment]
Summarizing: for every _pook.Mock_ there should be a _mocket.mockhttp.Entry_.

**Comment by h2non** [comment]
My point was about decoupling that part from `mocket`. 
The only relevant part is the `socket` mocking layer and HTTP parsing. 
Can't you get the rid of `Entry`?

**Comment by mindflayer** [comment]
Nope, it's the only way for Mocket to understand if it's something that can consume or something that should pass on the real _socket_.

**Comment by h2non** [comment]
I see. Ideally, `mocket` should separate concerns between socket mocking and mock matching, so then reusing parts of it would be easier.

However, assuming you don't want to change that, I jumped a bit into `mocket` code and I think you can easily overwrite the `Entry` limitation by changing this method:
https://github.com/mindflayer/python-mocket/blob/master/mocket/mocket.py#L163 

There you can switch and delegate that logic to the `pook` matcher engine.
Does it makes sense to you?

**Comment by mindflayer** [comment]
The HTTP mock relies a lot on Entry instances. I'll have a look at it, but I don't think I will be able to get rid of it, and I am also not sure it makes much sense at all.

**Comment by h2non** [comment]
Basically, you can parse the HTTP request buffer as you do now in order to compose the `Entry` object, but the different would be that instead of `Entry` you should compose `pook.Request`, which is the required known interface used during the mock matching phase.

From my perspective, the only useful part of `mocket` from `pook` perspective, would be the `socket` patch logic and HTTP request/response parser/builder.

I would considering giving it a try, and if you think it's introduces too much overhead on your side, we can workaround that with another approach.

The alternative I'm thinking could be optionally allowing the mock engine overwrites the mock constructor function at `pook` level, so you can do some sort of proxy/delegator pattern that would give you the ability to intercept HTTP mocks definitions via `pook` and build the required `Entry` instance at `mocket` level. That being said, I'm not quite happy with that approach and I think it would introduce overhead on your side anyway, and an unnecessary error-prone scenario.

**Comment by mindflayer** [comment]
Mocket is a socket mock framework, and the _Entry_ class is basically the definition of the protocol you want to mock, and implements its Request and Response classes.
I definitely prefer a simple hook where I can plug a function that creates an _Entry_ instance for each _pook.Mock_, because it's a fundamental concept of Mocket.
It's not a big deal, since we did not reinvent the wheel in both the implementations.

**Comment by h2non** [comment]
Got it. Let me explore that idea...

**Comment by h2non** [comment]
I'm thinking in a painless way of doing this in both ends... what about creating `Entry` objects at socket patched mock runtime? 

Since `pook.Engine` is injected into the `MockEngine`, you can retrieve all the `pook` mocks from there via `engine.mocks`:
https://github.com/h2non/pook/blob/master/pook/engine.py#L39-L40
Then you can inspect the `pook.Mock` instance and compose the equivalent `Entry` objects before triggering any `mocket` layer.

The limitation I see with a hook approach is that `pook.Mock` can be defined in a lazy way since it provides a fluent API, therefore `pook` doesn't exactly knows when a `pook.Mock` is fully defined.

Do you think my proposal sounds viable on your end?

**Comment by mindflayer** [comment]
It was my first approach, let me try it again following your hint.

**Comment by mindflayer** [comment]
I need to understand how to inspect  a _pook.Mock_. Looking into it while I am writing.

**Comment by h2non** [comment]
You would have to inspect MatchersEngine:
https://github.com/h2non/pook/blob/master/pook/mock.py#L119

Then, per each matcher there should be a `matcher.expectation` attribute that stores the mock value.

I know this is a bit tricky from your perspective, so I'm open to store mock data in pook.Request, so working with is would be more structured and simpler.

**Comment by mindflayer** [comment]
I have a _very simple_ working test that fails here:
`    assert mock.calls == 1`

**Comment by h2non** [comment]
Can you share some code?

**Comment by mindflayer** [comment]
`import pook
import requests
from pook.engine import MockEngine
from pook.interceptors.base import BaseInterceptor

from tests import urlparse, urlencode
from mocket.mocket import Mocket
from mocket.mockhttp import Entry


class MocketInterceptor(BaseInterceptor):
    def activate(self):
        Mocket.enable()

    def disable(self):
        Mocket.disable()


class MocketEngine(MockEngine):

    @staticmethod
    def set_query_parameter(url, key, value):
        """Given a URL, set or replace a query parameter and return the
        modified URL.

        >>> MocketEngine.set_query_parameter('http://example.com?foo=bar&biz=baz', 'foo', 'stuff')
        'http://example.com?foo=stuff&biz=baz'

        """
        scheme, netloc, path, query_string, fragment = urlparse.urlsplit(url)
        query_params = urlparse.parse_qs(query_string)

        query_params[key] = value
        new_query_string = urlencode(query_params, doseq=True)

        return urlparse.urlunsplit((scheme, netloc, path, new_query_string, fragment))

    def __init__(self, engine):
        # Store pook engine
        self.engine = engine
        # Store HTTP client interceptors
        self.interceptors = []
        # Self-register MocketInterceptor
        self.add_interceptor(MocketInterceptor)

    def activate(self):
        for mock in self.engine.mocks:
            response = mock._response

            method, url = [str(m) for m in mock.matchers]
            # query = request._query
            body = response._body
            status = response._status
            headers = response._headers

            # for k, v in query.items():
            #     MocketEngine.set_query_parameter(url, k, v)

            Entry.single_register(method, url, body, status, headers)

        super(MocketEngine, self).activate()


@pook.activate
def test_my_api():
    pook.set_mock_engine(MocketEngine)
    mock = pook.get('http://twitter.com/api/1/foobar', reply=404, response_json={'error': 'foo'})

    resp = requests.get('http://twitter.com/api/1/foobar')
    assert resp.status_code == 404
    assert resp.json() == {'error': 'foo'}
    assert mock.calls == 1
`

**Comment by tomas-fp** [comment]
I see. This is why I wanted to decouple as much as possible from `mocket` internals and just rely on the `socket` package patching.

The problem here is that `pook` matching engine is not triggered, the current implementation is an one-way flow only. The interceptors needs to trigger the matcher engine, such as:
https://github.com/h2non/pook/blob/master/pook/interceptors/urllib3.py#L90-L91

All the mock matching and counting logic is agnostic to the interceptor, but the interceptor is the entry point for this that notifies higher layers. Clearly, the matching engine is useless from the `mocket` perspective, since it would be already implemented there.

Honestly, I don't see how you can implement that in `mocket` without adding some hack somewhere. There should be a callback-like mechanism that should trigger the mock matching engine in order the trigger the required logic in `pook` side. 

I was thinking about registering a custom `Entry` object via mocket that inject a `pook` dependency. Then, we a request is matched and mocked out in `mocket`, a new logic should be added in `mocket` in order to infer if the `Entry` object was `pook` originated, and then trigger the matcher engine. It should be simple to achieve, but ideally I was thinking that integration should be frictionless in both ends without too much ad-hoc logic.

Does this sounds good to you? Do you have a better, alternative idea?

**Comment by mindflayer** [comment]
Sorry, I am really busy at the office since I'll be off for a week.
I do not understand why `mock._request` is not a valid `pook.request.Request` for each `self.engine.mocks`.

**Comment by h2non** [comment]
That's a good question: there's no concrete reason for that more than code simplicy and my lazyness.

The way I've designed HTTP mock matchers and the way I've integrated it from mock DSL perspective doesn't force me the proxy mock expectations in an equivalent `mock._request` instance, unless you explicitely use it.

In other words, that's something it can be relatively easy to fix, but adding complexibility in multiple areas. The current code is pretty simple because it doesn't have to delegate mock definitions in two places.

Also note that `mock._request` is a private member. It's not intended to be API level consumable by the user.

Would changing this help you?

**Comment by mindflayer** [comment]
I definitely think so, because `self.engine.match()` needs an instance of `pook.request.Request`.
My approach is extending the `Entry` class with something similar to the following lines:
```python
class MocketPookEntry(Entry):

    def can_handle(self, data):
        can_handle = super(MocketPookEntry, self).can_handle(data)

        if can_handle:
            pass  # call the self.engine.match() method
        return can_handle
```

**Comment by h2non** [comment]
I think in that case you won't need it. You can simply map the parsed HTTP intercepted request and construct the equivalent `pook.Request`, as it's already done via `pook.interceptors`:
https://github.com/h2non/pook/blob/master/pook/interceptors/aiohttp.py#L58-L67

**Comment by mindflayer** [comment]
The problem is having to introspect the `Mock` object, that seems very complicate.

**Comment by h2non** [comment]
In that case I think you can follow the same approach as before: iterate all the matchers and retrieve the `matcher.expectation` value. You would need to discern between each type of matcher, but they are [just a few](https://github.com/h2non/pook/tree/master/pook/matchers).

**Comment by mindflayer** [comment]
Of course I can, but I still think you should provide a better way to get a request instance like you do with your response. Everything would be way more consistent.

**Comment by h2non** [comment]
I did't forget about this, zI'm simply pretty busy this week. Hopefully, I eould be able to push something this weekend.

**Comment by mindflayer** [comment]
Take your time. Many thanks.

**Comment by h2non** [comment]
Finally, I had some free time :D 

We should be now in a position where this should no be a problem anymore. Note that the current implementation doesn't support custom, user-defined mock matchers, but that won't imply any limitation in the majority of cases.

For implementation details, see: https://github.com/h2non/pook/commit/2c6a16dba2ae17a877201c8f29bd963b46dadb8d

Those changes were shipped in `pook@v0.1.12`. 
Please, let me know how things evolve from your side :)

**Comment by mindflayer** [comment]
This is my solution:
```python
import pook
from pook.engine import MockEngine
from pook.interceptors.base import BaseInterceptor
import requests

from mocket.mocket import Mocket
from mocket.mockhttp import Entry, Response


class MocketPookEntry(Entry):
    pook_request = None
    pook_engine = None

    def can_handle(self, data):
        can_handle = super(MocketPookEntry, self).can_handle(data)

        if can_handle:
            self.pook_engine.match(self.pook_request)
        return can_handle

    @classmethod
    def single_register(cls, method, uri, body='', status=200, headers=None):
        entry = cls(uri, method, Response(body=body, status=status, headers=headers))
        Mocket.register(entry)
        return entry


class MocketInterceptor(BaseInterceptor):
    def activate(self):
        Mocket.disable()
        Mocket.enable()

    def disable(self):
        Mocket.disable()


class MocketEngine(MockEngine):

    def __init__(self, engine):
        # Store pook engine
        self.engine = engine
        # Store HTTP client interceptors
        self.interceptors = []
        # Self-register MocketInterceptor
        self.add_interceptor(MocketInterceptor)

    def activate(self):
        for mock in self.engine.mocks:

            request = mock._request
            method = request.method
            url = request.rawurl

            response = mock._response
            body = response._body
            status = response._status
            headers = response._headers

            entry = MocketPookEntry.single_register(method, url, body, status, headers)
            entry.pook_engine = self.engine
            entry.pook_request = request

        super(MocketEngine, self).activate()


@pook.activate
def test_my_api():
    pook.set_mock_engine(MocketEngine)
    mock = pook.get('http://twitter.com/api/1/foobar', body='{"a": 1}', headers={'content-type': 'application/json'}, reply=404, response_json={'error': 'foo'})

    resp = requests.get('http://twitter.com/api/1/foobar')
    assert resp.status_code == 404
    assert resp.json() == {'error': 'foo'}
    assert mock.calls == 1
```
The only thing still missing (something I've already experienced in the past) is `mock.calls` is not an integer, but your doc always compares it with an `int`. What's wrong with that?

And, in general, what do you think about this implementation?

**Comment by h2non** [comment]
> What's wrong with that?
That's a bug. It's fixed in `v0.1.13`.

> what do you think about this implementation?
That's was pretty much what I had in mind. I'm going to play with that a bit.

I'm open to add an example and update the documentation/README in order to inform the user that they can use `mocket` as underlying mock engine. 
What do you think? Do you want to provide an example code for that via PR?

**Comment by h2non** [comment]
Also, are you going to introduce that code as part of `mocket` code base, or are you going to release another package for this pook integration, such as `mocket-pook`?

**Comment by mindflayer** [comment]
That's a good question, I was thinking about something like `pip install mocket[pook]`.

**Comment by mindflayer** [comment]
Here it is:
https://github.com/mindflayer/python-mocket#works-well-with-others

**Comment by h2non** [comment]
Cool. It would be nice simplifying setup a bit. As an idea, the `MocketEngine` could be automatically registered in `pook` if both packages were installed together.

Also, you can simplify the example code like this:
```python
mock = pook.get(
    url,
    headers={'content-type': 'application/json'},
    reply=status,
    response_json=response_json,
    persist=True
)
```

I'm going to add an example and documentation for `mocket` integration as well in `pook` later.

**Comment by mindflayer** [comment]
Is this automation something you want to add to `pook`?

**Comment by h2non** [comment]
I'd prefer do not doing that automatically in `pook`, at least not until both packages integration is mature and working healthly. I was just thinking out loud.

**Comment by mindflayer** [comment]
Let me know if you experience something that requires my help.

**Comment by h2non** [comment]
`mocket` example code and documentation are now available in `pook`. For more details, see:
https://github.com/h2non/pook/commit/1d126ac2995d026dee81df3a5720f41b3e6198f6
https://github.com/h2non/pook/commit/2bf8ad0fa5a76ce0b71ca5a19fed502edaf5fc8a

Hopefully we can have some community feedback soon about how both libraries are working together.

I'm must say that I'm happy about the fact that we were eventually able get both packages integrated and working in convergence but also in a frictionless and not tighly-coupled way. 

This kind of scenarios is where open source ecosystem demostrates how effective and powerful it can be based on collaboration :)
